python/python/lance/sampler.py

```diff
@@ -67,12 +67,29 @@ def _efficient_sample(
     assert total_records > n
     chunk_size = total_records // max_takes
     chunk_sample_size = n // max_takes
+
+    num_sampled = 0
+
     for idx, i in enumerate(range(0, total_records, chunk_size)):
-        # Add more randomness within each chunk.
-        offset = i + np.random.randint(0, chunk_size - chunk_sample_size)
+        # If we have already sampled enough, break. This can happen if there
+        # is a remainder in the division.
+        if num_sampled >= n:
+            break
+        num_sampled += chunk_sample_size
+
+        # If we are at the last chunk, we may not have enough records to sample.
+        local_size = min(chunk_size, total_records - i)
+        local_sample_size = min(chunk_sample_size, local_size)
+
+        if local_sample_size < local_size:
+            # Add more randomness within each chunk, if there is room.
+            offset = i + np.random.randint(0, local_size - local_sample_size)
+        else:
+            offset = i
```

Comment:
How about if we simplify like this?
```suggestion
        # If we are at the last chunk, we may not have enough records to sample.
        local_sample_size = min(chunk_sample_size, total_records - i)

        num_sampled += local_sample_size
        offset = i + np.random.randint(0, local_sample_size)
```

Reply:
That looks like a totally different computation for `offset`? Offset needs to be constrained such that `offset + local_sample_size <= local_size`, otherwise we can get duplicate rows sampled.

Reply:
I'm thinking:
```python
        local_size = min(chunk_size, total_records - i)
        local_sample_size = min(chunk_sample_size, local_size)
```
equals:

```python
        local_sample_size = min(chunk_sample_size, min(chunk_size, total_records - i)) 
```
equals:

```python
        # because chunk_size > chunk_sample_size
        local_sample_size = min(chunk_sample_size, total_records - i)
```
`local_sample_size <= local_size` is always true because `local_sample_size = min(chunk_sample_size, local_size)` so we can just say `offset = i + np.random.randint(0, local_sample_size)`. Which I find more readable.
I hope I didn't miss something obvious :).

Reply:
I think the change is good either way, just looking to get rid of an extra variable.

Reply:
I still don't understand why

```python
offset = i + np.random.randint(0, local_sample_size)
```

is valid.

We are trying to find a random `offset` such that a window of size `local_sample_size` is inside of the window of size `local_size`. Hence the invariant I mentioned:

```
offset + local_sample_size <= local_size
```

```
 ◄────────────── local_size ──────────────►

 ┌──────┬────────────────┬────────────────┐
 │      │                │                │
 │i     │i + offset      │i + offset + lss│i + local_size
 │      │                │                │
 └──────┴────────────────┴────────────────┘

 ◄──┬───► ◄──────┬───────►
    │            │
 offset          │
         local_sample_size
```

Reply:
Oh I see! The offset would have to be:
```python
offset = i + np.random.randint(0, max(0, chunk_size - local_sample_size))
```
This is fine though :)